source/runtime.cpp

```diff
@@ -3182,9 +3228,27 @@ void reshade::runtime::reorder_techniques(std::vector<size_t> &&technique_indice
 void reshade::runtime::load_effects()
 {
 	// Reload preprocessor definitions from current preset before compiling
-	_preset_preprocessor_definitions.clear();
 	ini_file &preset = ini_file::load_cache(_current_preset_path);
-	preset.get({}, "PreprocessorDefinitions", _preset_preprocessor_definitions);
+
+	_preprocessor_definitions.erase(std::remove_if(_preprocessor_definitions.begin(), _preprocessor_definitions.end(),
+		[](const definition &definition) { return definition.scope == definition::scope_preset || definition.scope == definition::scope_effect; }), _preprocessor_definitions.end());
+
+	if (std::vector<std::pair<std::string, std::string >> preset_preprocessor_definitions;
+		preset.get({}, "PreprocessorDefinitions", preset_preprocessor_definitions))
+		for (std::pair<std::string, std::string> &value : preset_preprocessor_definitions)
+			_preprocessor_definitions.push_back(definition{ definition::scope_preset, "", std::move(value.first), std::move(value.second) });
+
+	if (std::vector<std::string> effect_names; preset.get(effect_names))
```

Comment:
This new `ini_file` method could be avoided as there already is a list of effect files (`effect_files`) build below, so could simply iterate over that to query the preset for each file.